File: Kalorie/Kalorie/Core/UseCases/GenerateFoodExportUseCase.swift
//
//  GenerateFoodExportUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 19.09.2026.
//

import ExportKit
import Foundation

protocol GenerateFoodExportUseCaseProtocol {
    func callAsFunction(from: Date, to: Date, format: FoodExportFormat, mealTypes: [MealTypeDomain]) async throws -> URL
}

struct GenerateFoodExportUseCase: GenerateFoodExportUseCaseProtocol {

    // MARK: - Properties

    private let fetchFoodsConsumedInRange: any FetchFoodsConsumedInRangeUseCaseProtocol
    private let reportFactory: FoodExportReportFactory
    private let directory: URL

    // MARK: - Init

    init(
        fetchFoodsConsumedInRange: any FetchFoodsConsumedInRangeUseCaseProtocol,
        reportFactory: FoodExportReportFactory = FoodExportReportFactory(),
        directory: URL = FileManager.default.temporaryDirectory
    ) {
        self.fetchFoodsConsumedInRange = fetchFoodsConsumedInRange
        self.reportFactory = reportFactory
        self.directory = directory
    }

    // MARK: - Functions

    func callAsFunction(from: Date, to: Date, format: FoodExportFormat, mealTypes: [MealTypeDomain]) async throws -> URL {
        let foods = try await fetchFoodsConsumedInRange(from: from, to: to)
        let report = reportFactory.makeReport(foods: foods, mealTypes: mealTypes, from: from, to: to)
        let data: Data
        switch format {
        case .pdf: data = NSDataBridgeKt.renderPdfData(report: report)
        case .xlsx: data = NSDataBridgeKt.renderXlsxData(report: report)
        }
        let url = directory.appendingPathComponent("Kalorie_\(isoDay(from))_\(isoDay(to)).\(format.fileExtension)")
        try data.write(to: url, options: .atomic)
        return url
    }

    // MARK: - Private

    private func isoDay(_ date: Date) -> String {
        date.formatted(Date.ISO8601FormatStyle(timeZone: .current).year().month().day())
    }
}

#if DEBUG
struct GenerateFoodExportUseCaseFake: GenerateFoodExportUseCaseProtocol {

    // MARK: - Properties

    var stubbedURL = URL(fileURLWithPath: "/dev/null")
    var stubbedError: Error?

    // MARK: - Functions

    func callAsFunction(from: Date, to: Date, format: FoodExportFormat, mealTypes: [MealTypeDomain]) async throws -> URL {
        if let stubbedError { throw stubbedError }
        return stubbedURL
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/GenerateFoodExportUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodExportFormat
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.exportkit.renderPdf
import antoni.kalorie.exportkit.renderXlsx
import java.io.File
import java.time.Instant
import java.time.ZoneId
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

interface GenerateFoodExportUseCaseProtocol {
    suspend operator fun invoke(from: Instant, to: Instant, format: FoodExportFormat, mealTypes: List<MealTypeDomain>): File
}

class GenerateFoodExportUseCase(
    private val fetchFoodsConsumedInRange: FetchFoodsConsumedInRangeUseCaseProtocol,
    private val reportFactory: FoodExportReportFactory,
    private val directory: File,
    private val zone: ZoneId = ZoneId.systemDefault(),
) : GenerateFoodExportUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(from: Instant, to: Instant, format: FoodExportFormat, mealTypes: List<MealTypeDomain>): File {
        val foods = fetchFoodsConsumedInRange(from, to)
        val report = reportFactory.makeReport(foods = foods, mealTypes = mealTypes, from = from, to = to)
        return withContext(Dispatchers.Default) {
            val data = when (format) {
                FoodExportFormat.PDF -> renderPdf(report)
                FoodExportFormat.XLSX -> renderXlsx(report)
            }
            directory.mkdirs()
            directory.listFiles()?.forEach { it.delete() }
            File(directory, "Kalorie_${isoDay(from)}_${isoDay(to)}.${format.fileExtension}").also { it.writeBytes(data) }
        }
    }

    // MARK: - Private

    private fun isoDay(instant: Instant): String = instant.atZone(zone).toLocalDate().toString()
}
